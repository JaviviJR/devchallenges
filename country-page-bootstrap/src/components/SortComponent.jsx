import PropTypes from 'prop-types';
import styles from './SortComponent.module.css';

function SortComponent({ sortByOptions, handler }) {
    
    const handleChange = (event) => {
        handler(event.target.value);
    }

    return (
        <div>
            <label htmlFor="sortby" className={`form-label ${styles.label1}`}>Sort by</label>
            <select name="sortby" className={`form-select ${styles.filterSelect}`} onChange={handleChange}>
                
                { sortByOptions.map(option => (
                    <option key={ option.value } value={ option.value }>{ option.label }</option>
                ))}

            </select>
        </div>
    );
}

export default SortComponent;

SortComponent.propTypes = {
    sortByOptions: PropTypes.arrayOf(PropTypes.shape({
        value: PropTypes.string.isRequired,
        label: PropTypes.string.isRequired
    })).isRequired,
    handler: PropTypes.func.isRequired
}