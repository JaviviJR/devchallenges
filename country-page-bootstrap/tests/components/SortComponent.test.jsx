import { render, screen } from "@testing-library/react"
import userEvent from '@testing-library/user-event'
import SortComponent from "../../src/components/SortComponent";

describe('testing Sort component', () => {
    const sortTest = [
        { value: 'sort1Value', label: 'sort1Label' },
        { value: 'sort2Value', label: 'sort2Label' }
    ];

    const handlerMock = jest.fn();

    beforeEach(() => {
        jest.clearAllMocks();
    });

    test('should match with snapshot', () => {
        const { container } = render(
            <SortComponent
                sortByOptions={ sortTest }
                handler={ handlerMock }
            />
        );

        expect(container).toMatchSnapshot();
    })

    test('should show all options', () => {
        render(
            <SortComponent
                sortByOptions={ sortTest }
                handler={ handlerMock }
            />
        );

        const combobox = screen.getByRole('combobox')
        const sort1 = screen.getByRole('option', {name: 'sort1Label'});
        const sort2 = screen.getByRole('option', {name: 'sort2Label'});

        expect( combobox ).toBeInTheDocument();
        expect( sort1 ).toBeInTheDocument();
        expect( sort2 ).toBeInTheDocument();

        expect(sort1.value).toBe('sort1Value');
        expect(sort2.value).toBe('sort2Value');
    });
    
    test('should handle click', async () => {
        const user = userEvent.setup()

        render(
            <SortComponent
                sortByOptions={ sortTest }
                handler={ handlerMock }
            />
        );

        const combobox = screen.getByRole('combobox')
        const sort1 = screen.getByRole('option', {name: 'sort1Label'});

        await user.click(combobox);
        await user.selectOptions(combobox, sort1);

        expect( handlerMock ).toHaveBeenLastCalledWith('sort1Value');
    });
})