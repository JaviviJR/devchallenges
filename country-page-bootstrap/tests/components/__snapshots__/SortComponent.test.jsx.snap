// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`testing Sort component should match with snapshot 1`] = `
<div>
  <div>
    <label
      class="form-label label1"
      for="sortby"
    >
      Sort by
    </label>
    <select
      class="form-select filterSelect"
      name="sortby"
    >
      <option
        value="sort1Value"
      >
        sort1Label
      </option>
      <option
        value="sort2Value"
      >
        sort2Label
      </option>
    </select>
  </div>
</div>
`;
